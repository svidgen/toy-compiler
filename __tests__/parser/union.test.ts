import {
	token,
	union
} from '../../src/parser';

describe('parser', () => {
	describe('Union', () => {
		describe('of non-optional string tokens', () => {
			const A = token('TOKEN_A', 'a');
			const B = token('TOKEN_B', 'b');
			const U = union('A_OR_B', [A, B]);

			it('can parse the first token', () => {
				const code = 'a';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('can parse the second token', () => {
				const code = 'b';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('fails if neither token is found', () => {
				const code = 'c';
				expect(() => U.parse({code})).toThrowErrorMatchingSnapshot();
			})
		});

		describe('of optional string tokens', () => {
			const A = token('TOKEN_A', 'a');
			const B = token('TOKEN_B', 'b');
			const U = union('A_OR_B', [A, B], true);

			it('can parse the first token', () => {
				const code = 'a';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('can parse the second token', () => {
				const code = 'b';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('returns `null` if neither token is found', () => {
				const code = 'c';
				const ast = U.parse({code});
				expect(ast).toEqual(null);
			})
		});

		describe('of non-optional regex tokens', () => {
			const A = token('TOKEN_A', /a/);
			const B = token('TOKEN_B', /b/);
			const U = union('A_OR_B', [A, B]);

			it('can parse the first token', () => {
				const code = 'a';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('can parse the second token', () => {
				const code = 'b';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('fails if neither token is found', () => {
				const code = 'c';
				expect(() => U.parse({code})).toThrowErrorMatchingSnapshot();
			});
		});

		describe('of optional regex tokens', () => {
			const A = token('TOKEN_A', /a/);
			const B = token('TOKEN_B', /b/);
			const U = union('A_OR_B', [A, B], true);

			it('can parse the first token', () => {
				const code = 'a';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('can parse the second token', () => {
				const code = 'b';
				const ast = U.parse({code});
				expect(ast).toMatchSnapshot();
			});

			it('returns `null` if neither token is found', () => {
				const code = 'c';
				const ast = U.parse({code});
				expect(ast).toEqual(null);
			});
		});
	});
});
