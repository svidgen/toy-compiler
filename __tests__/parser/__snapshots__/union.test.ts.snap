// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser Union of non-optional regex tokens can parse the first token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "a",
      "end": 1,
      "start": 0,
      "type": "TOKEN_A",
    },
  ],
  "code": "a",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of non-optional regex tokens can parse the second token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "b",
      "end": 1,
      "start": 0,
      "type": "TOKEN_B",
    },
  ],
  "code": "b",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of non-optional regex tokens fails if neither token is found 1`] = `
" expected a A_OR_B. Must be one of TOKEN_A, TOKEN_B.
line 0: c
        ^"
`;

exports[`parser Union of non-optional string tokens can parse the first token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "a",
      "end": 1,
      "start": 0,
      "type": "TOKEN_A",
    },
  ],
  "code": "a",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of non-optional string tokens can parse the second token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "b",
      "end": 1,
      "start": 0,
      "type": "TOKEN_B",
    },
  ],
  "code": "b",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of non-optional string tokens fails if neither token is found 1`] = `
" expected a A_OR_B. Must be one of TOKEN_A, TOKEN_B.
line 0: c
        ^"
`;

exports[`parser Union of optional regex tokens can parse the first token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "a",
      "end": 1,
      "start": 0,
      "type": "TOKEN_A",
    },
  ],
  "code": "a",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of optional regex tokens can parse the second token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "b",
      "end": 1,
      "start": 0,
      "type": "TOKEN_B",
    },
  ],
  "code": "b",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of optional string tokens can parse the first token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "a",
      "end": 1,
      "start": 0,
      "type": "TOKEN_A",
    },
  ],
  "code": "a",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;

exports[`parser Union of optional string tokens can parse the second token 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "b",
      "end": 1,
      "start": 0,
      "type": "TOKEN_B",
    },
  ],
  "code": "b",
  "end": 1,
  "start": 0,
  "type": "A_OR_B",
}
`;
